// @cdklabs/cdk-construct-analyzer/lib/cli/cli.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.cli = cli;
const chalk_1 = __importDefault(require("chalk"));
const yargs_1 = __importDefault(require("yargs"));
const helpers_1 = require("yargs/helpers");
const analyzer_1 = require("../library/analyzer");
function convertToStars(rating) {
    const fullStars = chalk_1.default.yellow('★'.repeat(rating));
    const emptyStars = chalk_1.default.gray('☆'.repeat(5 - rating));
    return fullStars + emptyStars;
}
/**
 * Converts signal names to Display Name format
 * Examples:
 * - "weeklyDownloads" -> "Weekly Downloads"
 * - "numberOfContributors_Maintenance" -> "Number Of Contributors - Maintenance"
 */
function convertToDisplayName(signalName) {
    return signalName
        .replace(/_/g, ' - ') // Convert underscore to dash with spaces
        .replace(/([A-Z])/g, ' $1') // Add space before capital letters everywhere
        .replace(/^./, str => str.toUpperCase()) // Capitalize first letter
        .replace(/\s+/g, ' ') // Clean up multiple spaces
        .trim();
}
function displayDetailsSignalInfo(signalScores, weights) {
    console.log(chalk_1.default.gray('\n---'));
    Object.entries(signalScores).forEach(([pillar, signals]) => {
        const pillarString = '\n=== ' + pillar + ' ===';
        console.log(chalk_1.default.bold.cyan(`${pillarString.padEnd(54)}`), chalk_1.default.bold('SCORE  WEIGHT'));
        Object.entries(signals).forEach(([signal, score]) => {
            const displayName = convertToDisplayName(signal);
            const dots = chalk_1.default.gray('.'.repeat(Math.max(1, 50 - displayName.length)));
            const stars = convertToStars(score);
            const signalWeight = weights[pillar][signal];
            console.log(`${chalk_1.default.dim('—')} ${displayName} ${dots} ${stars}    ${chalk_1.default.cyan(signalWeight)}`);
        });
    });
}
function cli() {
    const analyzer = new analyzer_1.ConstructAnalyzer();
    void (0, yargs_1.default)((0, helpers_1.hideBin)(process.argv))
        .command('$0 <package>', 'Usage: cdk-construct-analyzer <package> [options]', (yargsBuilder) => {
        return yargsBuilder
            .positional('package', {
            describe: 'NPM package name to analyze (e.g., "@aws-cdk/core")',
            type: 'string',
            demandOption: true,
        })
            .option('details', {
            alias: 'v',
            type: 'boolean',
            default: false,
            describe: 'Show detailed signal information',
        });
    }, async (argv) => {
        try {
            const result = await analyzer.analyzePackage(argv.package);
            const weights = result.signalWeights;
            console.log(chalk_1.default.bold('\nLIBRARY:'), chalk_1.default.blue(result.packageName));
            console.log(chalk_1.default.bold('VERSION:'), chalk_1.default.blue(result.version));
            // Color overall score based on value
            let scoreColor;
            if (result.totalScore >= 80) {
                scoreColor = chalk_1.default.green;
            }
            else if (result.totalScore >= 60) {
                scoreColor = chalk_1.default.yellow;
            }
            else {
                scoreColor = chalk_1.default.red;
            }
            console.log(chalk_1.default.bold('\nOVERALL SCORE:'), scoreColor(`${result.totalScore}/100`));
            console.log(chalk_1.default.gray('\n---'));
            console.log(chalk_1.default.bold('\nSUBSCORES'));
            Object.entries(result.pillarScores).forEach(([pillar, score]) => {
                // Color subscore based on value
                let pillarColor;
                if (score >= 80) {
                    pillarColor = chalk_1.default.green;
                }
                else if (score >= 60) {
                    pillarColor = chalk_1.default.yellow;
                }
                else {
                    pillarColor = chalk_1.default.red;
                }
                console.log(`  ${chalk_1.default.cyan(pillar.padEnd(12))}: ${pillarColor(score.toString().padStart(12) + '/100')}`);
            });
            // Only show detailed signal information if details flag is set
            if (argv.details) {
                displayDetailsSignalInfo(result.signalScores, weights);
            }
        }
        catch (error) {
            console.error(chalk_1.default.red('Error:'), error instanceof Error ? error.message : error);
            process.exit(1);
        }
    })
        .help()
        .argv;
}
