// @cdklabs/cdk-construct-analyzer/lib/library/analyzer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConstructAnalyzer = void 0;
const config_1 = require("./config");
const collect_1 = require("./data/collect");
class ConstructAnalyzer {
    constructor() {
        this.config = config_1.CONFIG;
    }
    async analyzePackage(packageName, weights) {
        const packageData = await (0, collect_1.collectPackageData)(packageName);
        const version = packageData.version;
        const { signalScores, pillarScores, totalScore } = await this.calculateSignalScores(packageData, weights);
        const normalizedPillarScores = this.normalizePillarScores(pillarScores, weights);
        const signalWeights = this.getSignalWeights(weights);
        return {
            packageName,
            version,
            totalScore,
            pillarScores: normalizedPillarScores,
            signalScores,
            signalWeights,
        };
    }
    async calculateSignalScores(packageData, weights) {
        var _a;
        const signalScores = {};
        const pillarScores = Object.fromEntries(Object.values(config_1.Pillar).map(pillar => [pillar, 0]));
        let totalWeightedSum = 0;
        let totalWeight = 0;
        for (const pillar of this.config.pillars) {
            for (const signal of pillar.signals) {
                const level = signal.benchmarks(packageData[signal.name]);
                const points = this.convertLevelToPoints(level, signal.name);
                const weight = weights?.[signal.name] ?? signal.defaultWeight;
                (signalScores[_a = pillar.name] ?? (signalScores[_a] = {}))[signal.name] = level ?? 1;
                pillarScores[pillar.name] = (pillarScores[pillar.name] || 0) + points * weight;
                totalWeightedSum += points * weight;
                totalWeight += weight;
            }
        }
        if (totalWeight != 100) {
            console.warn(`Warning: Signal weights sum to ${totalWeight} instead of 100. ` +
                'Weights should sum to 100 as it\'s universally understood and can be interpreted as percentages. ' +
                'Weights will be automatically normalized.');
        }
        const totalScore = totalWeight > 0 ? Math.round(totalWeightedSum / totalWeight) : 0;
        return { signalScores, pillarScores, totalScore };
    }
    convertLevelToPoints(level, signalName) {
        if (level == undefined) {
            console.warn(`Signal data not found: ${signalName}, assigning score of 0`);
            return 0;
        }
        return (level - 1) * 25;
    }
    normalizePillarScores(pillarScores, weights) {
        const normalizedScores = Object.fromEntries(Object.values(config_1.Pillar).map(pillar => [pillar, 0]));
        const pillarEntries = Object.entries(pillarScores);
        for (const [pillar, weightedSum] of pillarEntries) {
            const totalWeight = this.getTotalWeightForPillar(pillar, weights);
            const normalizedScore = totalWeight > 0 ? Math.min(100, weightedSum / totalWeight) : 0;
            normalizedScores[pillar] = Math.round(normalizedScore);
        }
        return normalizedScores;
    }
    getTotalWeightForPillar(pillarName, weights) {
        const pillar = this.config.pillars.find(p => p.name === pillarName);
        if (!pillar)
            return 0;
        return pillar.signals.reduce((sum, signal) => {
            const weight = weights?.[signal.name] ?? signal.defaultWeight;
            return sum + weight;
        }, 0);
    }
    /**
     * Extract signal weights from config in the same structure as signalScores
     * Uses custom weights when provided, otherwise falls back to default weights
     */
    getSignalWeights(weights) {
        var _a;
        const signalWeights = {};
        for (const pillar of this.config.pillars) {
            for (const signal of pillar.signals) {
                const weight = weights?.[signal.name] ?? signal.defaultWeight;
                (signalWeights[_a = pillar.name] ?? (signalWeights[_a] = {}))[signal.name] = weight;
            }
        }
        return signalWeights;
    }
}
exports.ConstructAnalyzer = ConstructAnalyzer;
