// @cdklabs/cdk-construct-analyzer/lib/library/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CONFIG = exports.Pillar = void 0;
const scoring_1 = require("./scoring");
/**
 * Pillar enum - single source of truth for pillar names
 */
var Pillar;
(function (Pillar) {
    Pillar["MAINTENANCE"] = "MAINTENANCE";
    Pillar["QUALITY"] = "QUALITY";
    Pillar["POPULARITY"] = "POPULARITY";
})(Pillar || (exports.Pillar = Pillar = {}));
/**
 * Main configuration object with all signals and their benchmarks
 */
exports.CONFIG = {
    pillars: [
        {
            name: Pillar.MAINTENANCE,
            description: 'Measures how actively maintained and updated the package is',
            signals: [
                {
                    name: 'timeToFirstResponse',
                    defaultWeight: 10,
                    description: 'Time to first response on issues',
                    benchmarks: (weeks) => (0, scoring_1.categorizeLowerIsBetter)([1, 4, 12, 52], weeks),
                },
                {
                    name: 'provenanceVerification',
                    defaultWeight: 10,
                    description: 'Ensures supply chain security through provenance verification',
                    benchmarks: (verified) => (0, scoring_1.categorizeByChecklist)({
                        versionVerified: { present: verified, value: 4 },
                    }),
                },
                {
                    name: 'releaseFrequency',
                    defaultWeight: 10,
                    description: 'Number of releases in the past year',
                    benchmarks: (releases) => (0, scoring_1.categorizeHigherIsBetter)([55, 34, 5, 1], releases),
                },
                {
                    name: 'numberOfContributors_Maintenance',
                    defaultWeight: 5,
                    description: 'Number of Contributors in the past year',
                    benchmarks: (contributors) => (0, scoring_1.categorizeHigherIsBetter)([8, 2, 1, 1], contributors),
                },
                {
                    name: 'numberOfFeatsAndFixes',
                    defaultWeight: 5,
                    description: 'Number of features and bug fixes contributed to the project in the past year',
                    benchmarks: (contributors) => (0, scoring_1.categorizeHigherIsBetter)([28, 10, 2, 1], contributors),
                },
                {
                    name: 'openIssuesRatio',
                    defaultWeight: 5,
                    description: 'Open issues / total issues',
                    benchmarks: (ratio) => (0, scoring_1.categorizeLowerIsBetter)([25, 50, 75, 90], ratio),
                    // ratio will be 100 if 0 total issues
                },
            ],
        },
        {
            name: Pillar.QUALITY,
            description: 'Measures the overall quality and reliability of the package',
            signals: [
                {
                    name: 'documentationCompleteness',
                    defaultWeight: 5,
                    description: 'Presence of README, API reference, and usage examples',
                    benchmarks: (docData) => (0, scoring_1.categorizeByChecklist)({
                        readme: { present: docData.hasReadme, value: 1 },
                        apiDocs: { present: docData.hasApiDocs, value: 1 },
                        oneExample: { present: docData.hasExample, value: 1 },
                        multipleExamples: { present: docData.multipleExamples, value: 1 },
                    }),
                },
                {
                    name: 'testsChecklist',
                    defaultWeight: 5,
                    description: 'Presence of unit tests and snapshot tests',
                    benchmarks: (testsData) => (0, scoring_1.categorizeByChecklist)({
                        unitTests: { present: testsData.hasUnitTests, value: 2 },
                        snapshotTests: { present: testsData.hasSnapshotTests, value: 2 },
                    }),
                },
                {
                    name: 'authorPackageCount',
                    defaultWeight: 5,
                    description: 'Highest package count among authors',
                    benchmarks: (packageCount) => (0, scoring_1.categorizeHigherIsBetter)([20, 11, 5, 2], packageCount),
                },
                {
                    name: 'releaseNotesIncludeFeatsAndFixes',
                    defaultWeight: 5,
                    description: 'Presence of features and fixes in release notes',
                    benchmarks: (releaseNotesData) => (0, scoring_1.categorizeByChecklist)({
                        hasFeats: { present: releaseNotesData.hasFeats, value: 2 },
                        hasFixes: { present: releaseNotesData.hasFixes, value: 2 },
                    }),
                },
                {
                    name: 'stableVersioning',
                    defaultWeight: 5,
                    description: 'Package version stability and deprecation status',
                    benchmarks: (versionData) => (0, scoring_1.categorizeByChecklist)({
                        isStableMajorVersion: { present: versionData.isStableMajorVersion, value: 2 },
                        hasMinorReleases: { present: versionData.hasMinorReleases, value: 1 },
                        deprecated: { present: versionData.isDeprecated, value: -4 },
                    }, 2), // Starting score of 2
                },
                {
                    name: 'multiLanguageSupport',
                    defaultWeight: 5,
                    description: 'Number of programming languages supported via jsii (excluding typescript)',
                    benchmarks: (languageCount) => (0, scoring_1.categorizeHigherIsBetter)([4, 3, 2, 1], languageCount),
                },
            ],
        },
        {
            name: Pillar.POPULARITY,
            description: 'Measures how widely adopted and used the package is',
            signals: [
                {
                    name: 'weeklyDownloads',
                    defaultWeight: 10,
                    description: 'Weekly download count from npm',
                    benchmarks: (downloads) => (0, scoring_1.categorizeHigherIsBetter)([2500, 251, 41, 6], downloads),
                },
                {
                    name: 'githubStars',
                    defaultWeight: 10,
                    description: 'GitHub repository stars',
                    benchmarks: (stars) => (0, scoring_1.categorizeHigherIsBetter)([638, 28, 4, 1], stars),
                },
                {
                    name: 'numberOfContributors_Popularity',
                    defaultWeight: 5,
                    description: 'Number of Contributors in the past year',
                    benchmarks: (contributors) => (0, scoring_1.categorizeHigherIsBetter)([8, 2, 1, 1], contributors),
                },
            ],
        },
    ],
};
